src/SIL.XForge.Scripture/ClientApp/src/xforge-common/notice.service.ts

```diff
@@ -0,0 +1,154 @@
+    private percentComplete: number;
+    private isProgressBarShown: boolean;
+    private isLoadingNotice: boolean;
+    private loadingMessage: string;
+
+    constructor() {
+        this.notices = [];
+        this.timers = {};
+        this.percentComplete = 0;
+        this.isProgressBarShown = false;
+        this.isLoadingNotice = false;
+    }
+
+    push(type: () => string, message: string, details?: string, cannotClose?: boolean, time?: number): string {
+        const id = UtilityService.uuid();   
+        let obj = new Subject<string>();
+
+        // Give a default auto-close time to success notifications of 4 seconds
+        if (!time && type() === this.SUCCESS()) time = 4 * 1000;
+        if (time) {
+            obj.pipe(debounceTime(time)).subscribe((noteId) => {
+                this.removeById(noteId);
+                this.timers[noteId] = null;
+            })
+            obj.next(id);
+            this.timers[id] = obj;
+        }
+        const notice = new Notice(type(), message, id, details, false);
+    
+        if (details) {
+          details = details.replace(/<p>/gm, '\n');
+          details = details.replace(/<pre>/gm, '\n');
+          details = details.replace(/<\/p>/gm, '\n');
+          details = details.replace(/<\/pre>/gm, '\n');
+          details = details.replace(/<[^>]+>/gm, ''); // remove HTML
+          details = details.replace(/\\\//g, '/');
+          notice.details = details;
+        }
+    
+        if (cannotClose) {
+          notice.cannotClose = true;
+        }
+    
+        this.notices.push(notice);
+        return id;
+      }
+
+    removeById(id: string): void {
+        this.remove(this.getIndexById(id));
+    }
+
+    remove(index: number): void {
+        if (index != undefined) {
+            this.notices.splice(index, 1);
+        }
+    }
+
+    get(): Notice[] {
+        return this.notices;
+    }
+
+    getLoadMessage(): string {
+        return this.loadingMessage;
+    }
+
+    setLoading(message: string): void {
+        this.loadingMessage = message;
+        this.isLoadingNotice = true;
+    }
+
+    getPercentComplete(): number {
+        return this.percentComplete;
+    }
+
+    setPercentComplete(percent: number): void {
+        this.percentComplete = percent;
+        this.isProgressBarShown = true;
+    }
+
+    cancelProgressBar(): void {
+        this.isProgressBarShown = false;
+    }
+
+    showProgressBar(): boolean {
+        return this.isProgressBarShown && this.percentComplete > 4;
+    }
+
+    cancelLoading(): void {
+        this.loadingMessage = '';
+        this.isLoadingNotice = false;
+    }
+
+    isLoading(): boolean {
+        return this.isLoadingNotice;
+    }
+
+    private getIndexById(id: string): number {
+        for (let i = 0; i < this.notices.length; i++) {
+            if (this.notices[i].id === id) {
+                return i;
+            }
+        }
+    }
+
+    ERROR(): string {
```

Comment:
I think these are historically functions because in JS you don't have classes and therefore can't declare constants on a class. I think these could even be **static readonly** properties instead of methods.